src/middlewared/middlewared/plugins/catalog/update.py

```diff
@@ -0,0 +1,145 @@
+                )],
+                max_length=60,
+            ),
+            register=True,
+        )
+
+    @private
+    def extend(self, data, context):
+        data.update({
+            'id': data['label'],
+            'location': os.path.join(
+                context['catalog_dir'], convert_repository_to_path(OFFICIAL_CATALOG_REPO, OFFICIAL_CATALOG_BRANCH)
+            ),
+        })
+        return data
+
+    @private
+    async def extend_context(self, rows, extra):
+        if await self.dataset_mounted():
+            catalog_dir = catalog_ds_path((await self.middleware.call('docker.config'))['dataset'])
+        else:
+            # FIXME: This can eat lots of memory if it's a large catalog
+            catalog_dir = TMP_IX_APPS_CATALOGS
+
+        return {
+            'catalog_dir': catalog_dir,
+        }
+
+    @private
+    async def dataset_mounted(self):
+        if docker_ds := (await self.middleware.call('docker.config'))['dataset']:
+            return bool(await self.middleware.call(
+                'filesystem.mount_info', [
+                    ['mount_source', '=', os.path.join(docker_ds, 'catalogs')], ['fs_type', '=', 'zfs'],
+                ],
+            ))
+
+        return False
+
+    @private
+    async def common_validation(self, schema, data):
+        verrors = ValidationErrors()
+        if not data['preferred_trains']:
+            verrors.add(
+                f'{schema}.preferred_trains',
+                'At least 1 preferred train must be specified.'
+            )
+        if (
+            await self.middleware.call('system.product_type') == 'SCALE_ENTERPRISE' and
+            OFFICIAL_ENTERPRISE_TRAIN not in data['preferred_trains']
+        ):
+            verrors.add(
+                f'{schema}.preferred_trains',
+                f'Enterprise systems must at least have {OFFICIAL_ENTERPRISE_TRAIN!r} train enabled'
+            )
+
+        verrors.check()
+
+    @accepts(
+        Dict(
+            'catalog_update',
+            List('preferred_trains'),
+            update=True
+        )
+    )
+    async def do_update(self, data):
+        """
+        Update catalog preferences.
+        """
+        await self.common_validation('catalog_update', data)
+
+        await self.middleware.call('datastore.update', self._config.datastore, OFFICIAL_LABEL, data)
+
+        return await self.config()
+
+    @private
+    async def update_train_for_enterprise(self):
+        catalog = await self.middleware.call('catalog.get_instance', OFFICIAL_LABEL)
+        if await self.middleware.call('system.product_type') == 'SCALE_ENTERPRISE':
+            preferred_trains = []
+            # Logic coming from here
+            # https://github.com/truenas/middleware/blob/e7f2b29b6ff8fadcc9fdd8d7f104cbbf5172fc5a/src/middlewared
+            # /middlewared/plugins/catalogs_linux/update.py#L341
```

Comment:
This comment makes a reference to a path that is being removed in this PR?? `catalogs_linux` is going away or am I misunderstanding something?

Reply:
The link will still remain active and can be accessed later on as well to see why we have this logic in place

Reply:
Oh derp, I see the URL has the old git commit hash :smile: